[Feat] 내 활동 바텀 시트 - 좋아요

aken-you opened the pull request:
# 관련 이슈 번호

#433 

# 설명

## 맵 모드를 반환하는 커스텀 훅

맵 페이지에서 router path에 따라 바텀시트 컨텐츠와 마커들이 달라집니다.
따라서 현재 맵 페이지의 컨텐츠가 무엇인지 반환하는 커스텀 훅을 만들었습니다.

- /map (동네 마킹): MAP 
- /map/place (이 장소 마킹): PLACE
- /map/my (내 마킹): MY_MARK
- /map/my-activity (내활동): MY_ACTIVITY

```typescript
type MapMode = "MAP" | "PLACE" | "MY_MARK" | "MY_ACTIVITY";

export const useMapMode = (): MapMode | null => {
  const { pathname } = useLocation();

  if (pathname === ROUTER_PATH.MAP) return "MAP";
  if (pathname === ROUTER_PATH.PLACE) return "PLACE";
  if (pathname === ROUTER_PATH.MY_MARK) return "MY_MARK";
  if (pathname === ROUTER_PATH.MY_ACTIVITY) return "MY_ACTIVITY";

  return null;
};
```

## [내 활동] 버튼

<img width="402" alt="스크린샷 2024-11-04 오후 3 54 54" src="https://github.com/user-attachments/assets/d94c39cb-e230-4945-815a-117456e17f6e">

맵 페이지의 왼쪽 상단에 있는 [내활동] 버튼을 클릭하면 url이 `/map/my-activity`로 바뀌고, 바텀 시트의 내용도 변경됩니다.

```
export const CollectionButton = () => {
  const navigate = useNavigate();
  const mode = useMapMode();
  const isCollectionActive = mode === "MY_ACTIVITY";

  return (
    <Button
      type="button"
      variant="outlined"
      colorType={isCollectionActive ? "primary" : "tertiary"}
      size="medium"
      className="shadow-custom-1 border-none"
      aria-label="좋아요를 눌렀거나 저장한 마킹들 나타내기"
      onClick={() => {
        navigate(ROUTER_PATH.MY_ACTIVITY);
      }}
    >
   // ...
```

## 내 활동 바텀 시트

MyActivityList 컴포넌트는 내 활동 바텀 시트 컴포넌트로 사용자가 누른 좋아요나 북마크를 보여줍니다.
현재는 좋아요 리스트만 구현한 상태이고 다음 작업 때 저장됨 기능을 구현할 예정입니다.

## 좋아요 취소하는 mutation에 optimistic update 적용

내 활동의 좋아요에서 좋아요를 취소하는 경우, 해당 마킹을 목록에서 제거해야 합니다.
우선 optimistic update를 통해 좋아요 취소한 마킹을 제거하고 좋아요 목록을 저장하고 있는 쿼리를 무효화합니다.

```typescript
export const useDeleteLikeMarking = () => {
  const queryClient = useQueryClient();
  const mode = useMapMode();

  return useMutation<unknown, Error, DeleteLikeMarkingRequest>({
    mutationFn: deleteLikeMarking,
    onSuccess: (_, { markingId }) => {
      if (mode === "MY_ACTIVITY") {
        queryClient.setQueryData<InfiniteData<GetMyLikedMarkingListResponse>>(
          ["myLikedMarkingList"],
          (oldData) => {
            if (!oldData) return oldData;

            const { pages } = oldData;
            const newPages = pages.map((page) => ({
              ...page,
              markings: page.markings.filter(
                (marking) => marking.markingId !== markingId,
              ),
            }));

            return {
              ...oldData,
              pages: newPages,
            };
          },
        );

        queryClient.invalidateQueries({
          queryKey: ["myLikedMarkingList"], // 내가 좋아요 누른 마킹 리스트가 저장된 쿼리
        });
      }
    },
  });
};
```

aken-you committed:
feat[#433] : 내가 좋아요 누른 마킹 데이터 요청하는 훅 생성

aken-you committed:
feat[#433] : 맵 url에 따라 맵 모드를 반환하는 훅 생성

- /map: 동네 마킹
- /map/place: 이 장소 관련 마킹
- /map/my: 내 마킹
- /map/my-activity: 내 활동

aken-you committed:
feat[#433] : 내 활동 바텀 시트 생성

aken-you committed:
feat[#433] : [내 활동] 버튼을 클릭하면 바텀시트가 내 활동으로 변경

aken-you committed:
feat[#433] : 내 활동 바텀 시트에서 좋아요 취소할 경우 optimistic update

aken-you committed:
feat[#433] : useMapMode 적용

yonghyeun commented on src/entities/marking/constants/index.ts:
```diff
@@ -1,6 +1,7 @@
-import { GetMarkingListRequest, GetUserMarkingListRequest } from "../api";
-import { GetBoundaryMarkerListRequest } from "../api/getBoundaryMarkerList";
+import type { GetMarkingListRequest, GetUserMarkingListRequest } from "../api";
+import type { GetBoundaryMarkerListRequest } from "../api/getBoundaryMarkerList";
```
index 에서 반환하도록 변경해야 할 거 같아요 !

yonghyeun commented on src/entities/marking/hooks/index.ts:
```diff
@@ -10,9 +9,8 @@ export const useGetMarkerList = (boundsParams: {
-
-  const isMyPage = pathname === ROUTER_PATH.MY_MARK;
+  const mapMode = useMapMode();
+  const isMyPage = mapMode === "MY_MARK";
```
my_mark 페이지에서는 useGetMyMakerList 만 호출하고 

아닌 페이지에선 useGetBoundaryMarkerList 만 호출하는 편은 어떨까요 ? 

useGetBoundaryMarkerList 훅에서 조건부로 처리하지 말고요

aken-you replied:
밑에 코멘트 남겼습니다!

yonghyeun commented on src/features/map/hooks/useMapMode.ts:
```diff
@@ -0,0 +1,15 @@
+export const useMapMode = (): MapMode | null => {
+  const { pathname } = useLocation();
+
+  if (pathname === ROUTER_PATH.MAP) return "MAP";
```
이 부분은 하나하나 하드코딩으로 조건부 처리하는 편보다 split 한 후에 객체에서 키로 반환을 하거나 그런 방법이 있을 수 있을 거 같아요 

아니면 경로에 따라 모드를 반환해야 하는 이유가 있을까요 ? 

핀 마커에서 직접적으로 api 요청을 받지 않고 props 로 데이터를 받는 편으로 하고 

라우팅 주소 별 페이지에 따라 다른 쿼리문을 날리는편이 어떨까요 ?

aken-you replied:
밑에 코멘트 남겼습니다!

yonghyeun commented on src/features/map/ui/mapControlButton.tsx:
```diff
@@ -88,10 +92,10 @@ const buttonBaseStyles = "border-none outline-none h-14 px-[.875rem]";
   const map = useMap();
 
-  const shouldShowMyMarking = pathname === ROUTER_PATH.MY_MARK;
+  const shouldShowMyMarking = mapMode === "MY_MARK";
```
이 기능은 NavLink 컴포넌트를 활용해보는건 어떨까요 ?

aken-you replied:
기존에 Button 컴포넌트를 사용하고 있어서 NavLink로 어떻게 구현할 수 있을까요?!

yonghyeun reviewed (commented):
Pin 마커 내부에서 api 를 호출하지 말고 리스폰스 값을 외부에서 props 로 내려 받는 편이 어떨까요 ? 

현재 useGetBoundaryMarkerList 는 여러 경로에 따라 `mapMode` 라는 값에 따라 다른 값을 반환하고 있는데 

차라리 mapMode 를 만들지 않고 경로 별로 페이지를 생성 한 후 

각 경로마다 서로 다른 독립적인 쿼리문을 뱉는 편이 어떨까요 

그 편이 더 관리하기 편해보입니다 

현재는 쿼리 문이 조건부적으로 응답값을 반환해야 해서 맵 모드라는 훅이 생긴 거 같은데 

이 문제는 라우팅 되는 경로 별로 서로 다른 쿼리문을 호출하는 편이 더 자연스러워 보이는 거 같습니다 

> 아니면 라우팅 경로별로 페이지를 만드는 것이 별로라면 차라리 동적 파라미터로 만들어두고 맵 모드를 쓰기 보다 useParams 로 하는 편은 어떨까 싶습니다

aken-you committed:
refactor[#433] : api 폴더에 있는 타입 import 주소 수정

aken-you commented:
>Pin 마커 내부에서 api 를 호출하지 말고 리스폰스 값을 외부에서 props 로 내려 받는 편이 어떨까요 ?
현재 useGetBoundaryMarkerList 는 여러 경로에 따라 mapMode 라는 값에 따라 다른 값을 반환하고 있는데
차라리 mapMode 를 만들지 않고 경로 별로 페이지를 생성 한 후
각 경로마다 서로 다른 독립적인 쿼리문을 뱉는 편이 어떨까요

useGetBoundaryMarkerList가 아니라 useGetMarkerList 말씀해주시는 것 같아요! 

우선 맵 페이지의 구조를 보시면 PinMarker는 MapMarkerWidget에 있습니다.

<img width="810" alt="스크린샷 2024-11-04 오후 10 48 53" src="https://github.com/user-attachments/assets/44a347db-e44c-4553-a82f-9b32460d12c5">

PinMarker에서 Props로 markerlist를 받게 되면 MapMarkerWidget에서 모드에 따라 다른 마커 리스트를 호출해야 하는 건 동일합니다.
그래서 props로 받기 위해서 PinMarker의 위치를 각 페이지의 바텀 시트로 옮겨야 합니다.
그렇게 되면 맵과 관련된 바텀시트에 PinMarker를 포함해야 합니다.

선택지는 2개인 듯 합니다.
1. PinMarker를 MapMarkerWidget에 두되 불러올 마킹 리스트를 분기 처리하기
2. 각 바텀시트에 PinMarker를 포함하기

>현재는 쿼리 문이 조건부적으로 응답값을 반환해야 해서 맵 모드라는 훅이 생긴 거 같은데
이 문제는 라우팅 되는 경로 별로 서로 다른 쿼리문을 호출하는 편이 더 자연스러워 보이는 거 같습니다

>이 부분은 하나하나 하드코딩으로 조건부 처리하는 편보다 split 한 후에 객체에서 키로 반환을 하거나 그런 방법이 있을 수 있을 거 같아요
아니면 경로에 따라 모드를 반환해야 하는 이유가 있을까요 ?
핀 마커에서 직접적으로 api 요청을 받지 않고 props 로 데이터를 받는 편으로 하고
라우팅 주소 별 페이지에 따라 다른 쿼리문을 날리는편이 어떨까요 ?

경로에 따라 받아야 하는 마커 데이터가 다르기 때문에 맵 모드 훅을 만든 게 맞습니다.
제시해주신 방법으로 하기 위해 PinMarker의 위치를 옮겨야 할 것 같습니다.

aken-you committed:
refactor[#433] : 동네 마킹 페이지 컴포넌트 생성

- 기존에 MapPage에 있던 코드들은 MapLayout으로 이동
- MapMarkerWidget에 있던 PinMarker, MultiplePinMarker, ClusterMarker 삭제

aken-you committed:
refactor[#433] : 내 마킹 페이지 컴포넌트 생성

aken-you committed:
refactor[#433] : 이 장소 마킹 페이지 컴포넌트 생성

aken-you committed:
refactor[#433] : useGetMarkerList 삭제

aken-you committed:
refactor[#433] : MapMarkerWidget에 있던 PinMarker, MultiplePinMarker, ClusterMarker 삭제

aken-you commented:
저희끼리 회의했던대로 각 바텀시트에 PinMarker를 포함하기로 했습니다.

## 마커에 표시될 핀 위치 이동

이전에는 MapMarkerWidget에 핀 컴포넌트들이 다 있었지만, 삭제하고 각 맵 페이지 컴포넌트에 포함하였습니다.
MapPage에 있던 코드를 MapLayout에 옮겨두었고, MapPage에는 동네 마킹에 관련된 컨텐츠가 있습니다.

하지만, 내 활동 경로에는 아직 element가 MyActivityList인데요.
아직 좋아요, 저장했던 마커들만 반환하는 api가 없습니다. 내일 회의 때 이야기를 나눠보기로 해서 결정되면 그 때 수정하려고 합니다.

```typescript
export const router = createBrowserRouter([
  {
    path: ROUTER_PATH.MAIN,
    element: <AppProviderLayout />,
    children: [
      {
        index: true,
        element: <MainPage />, // 탐색
      },
      {
        path: ROUTER_PATH.MAP,
        element: <MapLayout />, // 지도
        children: [
          {
            index: true,
            element: <MapPage />,
          },
          {
            path: ROUTER_PATH.PLACE,
            element: <PlaceMarkingPage />,
          },
          {
            path: ROUTER_PATH.MY_MARK,
            element: <MyMarkingPage />,
          },
          {
            path: ROUTER_PATH.MY_ACTIVITY,
            element: <MyActivityList />,
          },
        ],
      },
```

이 변경으로 인해 useGetMarkerList 훅과 MapMarkerWidget에 있던 PinMarker, MultiplePinMarker, ClusterMarker를 삭제했습니다.
페이지에서 핀을 정의하니 훨씬 깔끔해져서 다행입니다 👍

aken-you committed:
refactor[#448] : useMapMode에서 mapMode 객체 생성

aken-you committed:
Merge branch 'dev' into feat/#433-like
